### PR TITLE
#5 Search

### PR DESCRIPTION
**Pull request by 3rdvision**
| - | - |
| --- | --- |
| Issue | https://github.com/ripe-tech/ripe-id-mobile/issues/5 |
| Decisions | - Add search component <br> - Add `placeholderTextColor` to textare component <br> - Add select mode prop to allow selecting from either "dropdown" or "modal" modes. |
| Animated GIF | Peek 2021-07-07 18-08 |

### CODE REVIEW

File: react/components/molecules/search/search.js [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+import { ButtonIcon, TextArea } from "../../atoms";
+
+export class Search extends PureComponent {
+    static get propTypes() {
+        return {
+            value: PropTypes.string,
+            placeholder: PropTypes.string,
+            buttonProps: PropTypes.object,
+            onFocus: PropTypes.func,
+            onValue: PropTypes.func,
```
Comment (BeeMargarida) [inline react/components/molecules/search/search.js:14]: This is nitpicking, but maybe move onValue above onFocus, so that onFocus and onBlur stay together, since they are opposites

File: react/components/molecules/search/search.js [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        super(props);
+
+        this.state = {
+            value: this.props.value,
+            buttonsVisible: true,
+            focused: false
+        };
+    }
+
+    setValue = (value, callback = () => {}) => this.setState({ value: value }, callback);
```
Comment (BeeMargarida) [inline react/components/molecules/search/search.js:46]: Will this be used? Or is it dead code?

File: react/components/molecules/search/search.js [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+    focus = () => {
+        this.textAreaComponent.focus();
+    };
+
+    blur = () => {
+        this.textAreaComponent.blur();
+    };
+
+    sendMessage = () => {
```
Comment (BeeMargarida) [inline react/components/molecules/search/search.js:56]: dead code?

File: react/components/molecules/search/search.js [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        this.props.onSendMessage(this.state.value);
+    };
+
+    clear = () => {
+        const emptyValue = "";
+        this.setState({ value: emptyValue }, () => this.props.onValue(emptyValue));
+        this.focus();
+        this.props.onClear();
```
Comment (BeeMargarida) [inline react/components/molecules/search/search.js:64]: Maybe add this call to the setState callback as well?

File: react/components/molecules/search/search.js [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    blur = () => {
+        this.textAreaComponent.blur();
+    };
+
+    sendMessage = () => {
+        this.props.onSendMessage(this.state.value);
+    };
+
+    clear = () => {
+        const emptyValue = "";
```
Comment (BeeMargarida) [inline react/components/molecules/search/search.js:61]: Maybe just use `""` instead of assigning it to a variable What do you think?

File: react/components/molecules/search/search.js [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+            onSubmit: () => {},
+            style: {}
+        };
+    }
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            value: this.props.value,
```
Comment (BeeMargarida) [inline react/components/molecules/search/search.js:40]: Even though this brings no problem, in the majority of the components in react we are using `valueData`, similar to vue. This helps with code readability, since its very ease to confuse prop "value" it state "value"

File: react/components/molecules/select/select.js [hunk 7]
```diff
@@ -14,6 +14,7 @@ export class Select extends mix(PureComponent).with(IdentifiableMixin) {
             options: PropTypes.array,
             value: PropTypes.any,
             placeholder: PropTypes.string,
+            mode: PropTypes.string,
```
Comment (BeeMargarida) [inline react/components/molecules/select/select.js:17]: These changes are not related to this PR, right?

File: react/components/molecules/search/search.js [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        });
+    };
+
+    onTextAreaSubmit = () => {
+        this.setState({ focused: false }, () => {
+            this.props.onSubmit();
+        });
+    };
+
+    onButtonPress = () => {
```
Comment (BeeMargarida) [inline react/components/molecules/search/search.js:90]: Is this normal in applications? I don't remember having to tap de search icon to submit a search, normally the search is made each time we type or remove a character. Its does not make sense to submit an empty text as well. The `search` in  `ripe-components-vue` does not have this behaviour as well.

I think you can remove this (the onSubmit behaviour), and maybe even use `Input` instead of `TextArea` (although I don't know if it behaves exactly the same). What do you think?

File: react/components/molecules/search/search.js [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        flex: 1,
+        paddingBottom: 0,
+        paddingTop: 0,
+        borderRadius: 6,
+        backgroundColor: "#ffffff"
+    },
+    button: {
+        position: "absolute",
+        backgroundColor: "transparent",
+        top: 13,
```
Comment (BeeMargarida) [inline react/components/molecules/search/search.js:150]: A possible fix for iOS centering 
image

```suggestion
        top: Platform.OS === "ios" ? 8 : 13,
```

File: react/components/molecules/search/search.js (hunk 8 above)
Comment (3rdvision) [reply react/components/molecules/search/search.js:90]: we need this to clear the text when clicking the `close` icon

File: react/components/molecules/card/card.js [hunk 10]
```diff
@@ -9,13 +9,13 @@ import { baseStyles } from "../../../util";
...
     static get propTypes() {
         return {
-            avatarURL: PropTypes.String,
-            icon: PropTypes.String,
-            iconColor: PropTypes.String,
-            iconFill: PropTypes.String,
-            title: PropTypes.String,
-            text: PropTypes.String,
-            subText: PropTypes.String,
+            avatarURL: PropTypes.string,
```
Comment (BeeMargarida) [inline react/components/molecules/card/card.js:12]: Move this fixes to the card PR, since they are not related to this PR

File: react/components/molecules/search/search.js [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+        paddingBottom: 0,
+        paddingTop: 0,
+        borderRadius: 6,
+        minHeight: 40,
+        backgroundColor: "#ffffff"
+    },
+    button: {
+        position: "absolute",
+        backgroundColor: "transparent",
+        top: Platform.OS === "ios" ? 8 : 13,
```
Comment (BeeMargarida) [inline react/components/molecules/search/search.js:138]: well, something must have changed, now it can be `top: 8` for both, in android is currently looking like this 
image

**Review by BeeMargarida — COMMENTED** [review]
Almost there, looking good!

**Review by BeeMargarida — COMMENTED** [review]
Just one small thing, I think the padding on the left needs to be a little bit bigger:
| - | - |
| --- | --- |
| Current | Design |
| image | image |

File: react/components/molecules/search/search.js [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+        flexDirection: "row",
+        alignItems: "center",
+        position: "relative"
+    },
+    input: {
+        flex: 1,
+        color: "#223645",
+        paddingBottom: 0,
+        fontSize: 14,
+        fontFamily: "Roboto-Medium",
```
Comment (BeeMargarida) [inline react/components/molecules/search/search.js:131]: Use baseStyle.FONT for now

Comment (3rdvision) [reply react/components/molecules/search/search.js:131]: We can now use Roboto font as, both PRs with that font got merged

https://github.com/ripe-tech/ripe-id-mobile/pull/14
https://github.com/ripe-tech/ripe-components-react-native-storybook/pull/24

Comment (BeeMargarida) [reply react/components/molecules/search/search.js:131]: I understand, but we will not use it like that, we use it as baseStyle.FONT. In the future, the FONT constant will be changed to point to the Roboto font, changing both ripe-id and ripe-robin apps

**Review by BeeMargarida — APPROVED** [review]

**Review by gcandal — APPROVED** [review]

**Review by joamag — APPROVED** [review]
